src/librbd/ImageCtx.h

```diff
@@ -202,7 +202,7 @@ namespace librbd {
 
     typedef boost::lockfree::queue<
       io::AioCompletion*,
-      boost::lockfree::allocator<ceph::allocator<void>>> Completions;
+      boost::lockfree::allocator<ceph::allocator<void*>>> Completions;
```

Comment:
please note `std::allocator<void>` has an explicit specialization before C++20. i think the reason why we pass our own allocator to `boost::lockfree::allocator` for the allocator type in 7f6685383dfdab1b685c9c4df609feb7fdc97a91 is to address http://tracker.ceph.com/issues/39703 to prevent us from using the missing `aligned_alloc()` in tcmalloc. if no customized allocator is specified, boost uses `std::allocator<node>` for allocating the internal nodes of `boost::lockfree::queue`. otherwise, the user-specified one is used by rebinding to it. since `std::allocator<void>` does not have any typedefs in it. it's a perfect fit in this case. we always rebind it here.

but the element type of `std::allocator<void*>`  is `void*`, which does not sound *correct* to me.

i am proposing #41340 as an alternative.

Reply:
Yeah, `std::allocator<void*>` seems weird.  I guess it depends on how you choose to think about the allocator: whether it allocates raw memory or actual objects.  If the latter, `std::allocator<void>` is also meaningless...

Anyway, I agree that getting rid of `ceph::allocator` if we can is better.

Reply:
Hi @tchaikov

I don't feel really like argueing on this with you. You know a lot more about the intricasies of C++ than I do.
I'm just restating the obvious where `libc++` claims that even in C++17 it an not wanted construction.
And not sure if that will not apply voor the `tcmalloc` one in the near future as well.
I looked it up, and then I/we get into very pricipal C++ standards commitee discussion. Don't really want to go there.

`<void*>` seemed to most harmless other type to apply. 
Note that the FreeBSD port seems to take the `std::allocator` shortcut in `src/common/allocator.h`
But I'm not sure that should be the case.

Looking at the alternative PR you just removed the bugfix of http://tracker.ceph.com/issues/39703. So I assume that the chance of running into `tcmalloc < 2.6.2` are slim to none.
But if you fix works, I'm happy to take it.
I'll run it thru Jenkins this weekend to make sure.

Reply:
> 
> 
> Yeah, `std::allocator<void*>` seems weird. I guess it depends on how you choose to think about the allocator: whether it allocates raw memory or actual objects. If the latter, `std::allocator<void>` is also meaningless...
> 
> Anyway, I agree that getting rid of `ceph::allocator` if we can is better.

This paper:
[http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0174r0#2.4](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0174r0#2.4)
is more or less where there are hints of not allowing `allocator<void>`. And I parafrase: "because it is not a true allocator"